### ISSUE TITLE
Revisions to identifiers & relationships

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/concerto#181]
(This is a place-holder issue, need more details and explanations).

Current status:
- Identifiers can be and have to be declared on transactions and events, but no other classes
- Relationships are distinguished from simple `has a` properties of a class, and have to be to an identifiable class
- The relationship can be to a resource or to an inline object

Some changes in `release-1.0`:
- Any class can have an identifier
- Relationship can be to any class which has an identifier
- Internally identifiers are both kept as a special `$identifier` property and the user-defined property declared in the `identified by` part of a class declaration

Some further possible changes:
- Distinguish between internal identifiers (necessary for relationships) and user-defined properties

**Comment by jeromesimeon** [comment]
_Notes from discussion (@dselman @mttrbrts @jeromesimeon)_

## User stories

- When creating a model, one wants to distinguish "inline" data (e.g., company containing an address) from "relationship" data (e.g., purchase order referencing a company)
- When creating a model, one wants to distinguish external identifiers (e.g., SSN, DB key) from identifiers used for the sole purpose of creating a relationship
- When editing a contract, one wants to simply refer to the relationship e.g., party, but just want to edit the relationship identifier

## Current Issues

- When creating writing contract logic, one does not want to have to invent identifiers (See #...)
- Current "hack" that identifiers from the base model are hidden from logic does not apply now that systems models are gone
- Unhydrated relationships are unsound (Ergo compiler assumes the relationship is "hydrated")

## Technical proposal

1. New relationship / identifier model

   1. concepts can be identified or not
   2. a concept `C` can be used through a relationship (`--> C name`) iff it is identified
   3. a concept `C` can be used inline (`o C name`) iff it is not identified
   4. identifiers for identified concepts can be managed by the system (declared as `identified` and serialized as property `$identifier`) or by the users (declared as `identified by myId` and serialized as property `myId`)
   5. identifiers managed by the system are created through a `generate` function parameter whose signature is from object to a `String` identifier.
      The function should be complete and deterministic. A default function is provided which returns a stable hash based on the fields in the object.
      identifiers managed by the system do not appear (cannot be accessed or specified) in the contract logic.
   6. identifiers managed by the user have to be provided in the instance objects and are not automatically created.
      identifiers managed by the system always appear (can be accessed and have to be specified) in the contract logic.
   7. asset/participant/event/transaction are special concepts that are always identified

2. Resolving Relationships

   1. resolve allows users to specify how resources are "hydrated" i.e., resolved into their full value for processing
      resolve is a function from resources URIs to objects of the appropriate class.
      ```
      { "$class": "Order", "seller": "resource:org.accordproject.Organization#partyA" }
      --> resolve (may lookup in file/registry/DB/Web service)
      { "$class": "Order", "seller": { "$class: "org.accordproject.Organization", "$identifier": "partyA", "name": "ACME" } }
      ```
   2. unresolve allows users to specify how the result of processing should be "split" into resources
      ```
      { "$class": "Delivery", "recipient": { "$class: "org.accordproject.Organization", "$identifier": "partyB", "name": "FOOBAR" } }
      --> unresolve (may store in file/registry/DB/Web service)
      { "$class": "Delivery", "recipient": "resource:org.accordproject.Organization#partyB" }
      ```
   3. resolve/unresolve are added as parameters to Cicero API (for parse, draft & execute)
   4. Cicero CLI adds a `--registry` parameter containing resources, which is used by default `resolve` and `unresolve`
      `cicero parse --registry registry.json  ...`

3. Revision to Cicero models

New Cicero models are adjusted and take advantage of the new model

- Most redundant/user controlled identifiers now use system controlled identifiers (removing `transactionId` `eventId` `contractId` `clauseId` `obligationId`)
- Parties retain a user-controlled `partyId`

**Contract model**

```
namespace org.accordproject.contract

/**
 * Contract Data
 * -- Describes the structure of contracts and clauses
 */

/* A party to a contract */
participant Party identified by partyId {
  o String partyId
}

/* A contract is a asset -- This contains the contract data */
abstract asset Contract {
  --> Party[] parties optional
}

/* A clause is an asset -- This contains the clause data */
abstract asset Clause {
}
```

**Runtime model**

```
namespace org.accordproject.runtime

import org.accordproject.contract.Party from https://release-1-0--accordproject-models.netlify.com/accordproject/contract.cto
import org.accordproject.contract.ContractData from https://release-1-0--accordproject-models.netlify.com/accordproject/contract.cto

/**
 * Runtime API
 * -- Describes input and output of calls to a contract's clause
 */

/* A request is a transaction */
transaction Request {
}

/* A response is a transaction */
transaction Response {
}

/* An Error is a transaction */
abstract transaction ErrorResponse {
}

/* An event that represents an obligation that needs to be fulfilled */
abstract event Obligation {
  /* A back reference to the governing contract that emitted this obligation */
  --> Contract contract

  /* The party that is obligated */
  --> Party promisor

  /* The party that receives the performance */
  --> Party promisee

  /* The time before which the obligation is fulfilled */
  o DateTime deadline optional
}

/* A contract state is an asset -- The runtime state of the contract */
asset State {
	--> Obligation[] obligations // Should we allow for a standard way to store obligations on the contract in the state?
}
```

**Comment by dselman** [comment]
- [ ] Update peg grammar to allow `identified by` for concepts
- [ ] Update peg grammar to allow `identified` for concepts/assets/participants/transaction/events
- [] Update `ClassDeclaration` to create a new system field `$identifier` for `identified` classes
- [] Merge `Concept` and `Resource` in the model
- [] Update `Factory` to allow concepts to be created with an identifier

**Comment by dselman** [comment]
Changes have been merged into the release-1.0 branch.
